### PR TITLE
Data imports

### LINKED ISSUE #5130: Partner Data Import Bot (ol-home0 Docker Service)
Part of #2034

### Stakeholders
@BharatKalluri

### PR DESCRIPTION
**Pull request by mekarpeles**
This is a followup of #5583 which should be **merged first** which updates the schema for the `import_items` table.

Closes first half of #5130

Allows manage-imports.py to add `{"data": {...}}` records in addition to `{"ia_id": "..."}`records to the https://openlibrary.org/admin/imports pipeline.

:warning:  **Please squash**

### Testing

**`tty1`** 
```
ssh -A ol-home0.us.archive.org
# Monitor the ImportBot container (w/ old code) to make sure it's not
# the one attempting to import records -- we want to manually start another
# `import-all` service from the container directly after we enqueue records
# so we can be confident that our new code is performing the imports:
sudo docker logs -f --tail 100 openlibrary_importbot_1
```

**`tty2`**
```
# Connect to db
ssh -A ol-db1.us.archive.org
sudo -upostgres psql
# Delete old records you want to re-import, e.g:
delete from import_item where data is NOT NULL;
# Confirm only the records you expect are in the db
select * from import_item where data is NOT NULL;
```

**`tty3`**: 
```
ssh -A ol-home0.us.archive.org
# Connect to the cron container
sudo docker exec -uroot -it openlibrary_cron-jobs_1 bash
# rm the statefile (saving our progress processing & queuing data)
rm  /1/var/tmp/imports/2021-08/Bibliographic/2021-08-04/import.log
# re-run the test.py script to queue up a test batch of imports
python3 /olsystem/bin/bwb_etl/test.py /1/var/tmp/imports/2021-08/Bibliographic/*/bettworldbks*
```

**`tty4`**
```
ssh -A ol-home0.us.archive.org
# Connect to the ImportBot container
sudo docker exec -it -uroot openlibrary_importbot_1 bash
# Run the continuous import process
scripts/manage-imports.py --config /olsystem/etc/openlibrary.yml import-all
```

Go back to **`tty2`**
```
# ... Re-run this commend *after* you've enqueued items using test.py
select * from import_item where data is NOT NULL;
```

### CODE REVIEW

File: scripts/manage-imports.py [hunk 1]
```diff
@@ -99,6 +104,26 @@ def import_ocaids(*ocaids, **kwargs):
...
+    batch.add_items(items, ia_items=False)
+
+def import_data(args):
+    # [x] Testing imports work to new ol.import_data json data api
+    data = args[0]
+    ol = get_ol()
+    r = ol.import_data(data)
+    print(r)
+    return r
+
```
Comment (mekarpeles) [inline scripts/manage-imports.py:126]: rm testing code
```suggestion
```

File: scripts/manage-imports.py [hunk 2]
```diff
@@ -204,6 +228,10 @@ def main():
         return add_items(args)
     elif cmd == "add-new-scans":
         return add_new_scans(args)
+    elif cmd == "add-data":
+        return add_data(args)
+    elif cmd == "import-data":
+        return import_data(args)
```
Comment (mekarpeles) [inline scripts/manage-imports.py:214]: rm testing code
```suggestion
```

File: openlibrary/core/imports.py [hunk 3]
```diff
@@ -32,18 +33,48 @@ def load_items(self, filename):
...
+            try:
+                # Upgrads psql and use `INSERT OR IGNORE`
+                # otherwise it will fail on UNIQUE `data`
+                # https://stackoverflow.com/questions/1009584
+                db.get_db().multiple_insert("import_item", values)
+            except:
+                for value in values:
+                    try:
+                        db.get_db().insert("import_item", value)
+                        print("success")
```
Comment (mekarpeles) [inline openlibrary/core/imports.py:75]:
```suggestion
```

File: openlibrary/core/imports.py [hunk 4]
```diff
@@ -32,18 +33,48 @@ def load_items(self, filename):
...
+                    except:
+                        print("fail")
```
Comment (mekarpeles) [inline openlibrary/core/imports.py:77]:
```suggestion
```

File: openlibrary/core/imports.py [hunk 5]
```diff
@@ -32,18 +33,46 @@ def load_items(self, filename):
...
+        if ia_items:
+            items = self.dedupe_ia_items(items)
+
         if items:
-            values = [dict(batch_id=self.id, ia_id=item) for item in items]
-            db.get_db().multiple_insert("import_item", values)
+            # Either create a reference to an IA id which will be loaded
+            # from Archive.org metadata, or provide json data book record
+            # which will be loaded directly into the OL catalog
+            record = lambda item: {'ia_id': item} if ia_items else {'data': json.dumps(item)}
```
Comment (mekarpeles) [inline openlibrary/core/imports.py:64]:
```suggestion
            def record(item):
                return {'ia_id': item} if ia_items else {'data': json.dumps(item)}
```

File: openlibrary/core/imports.py [hunk 6]
```diff
@@ -32,18 +33,48 @@ def load_items(self, filename):
...
+                # Upgrads psql and use `INSERT OR IGNORE`
+                # otherwise it will fail on UNIQUE `data`
+                # https://stackoverflow.com/questions/1009584
+                db.get_db().multiple_insert("import_item", values)
+            except:
+                for value in values:
+                    try:
+                        db.get_db().insert("import_item", value)
+                        print("success")
+                    except:
```
Comment (mekarpeles) [inline openlibrary/core/imports.py:75]:
```suggestion
                    except Exception:
```

File: openlibrary/core/imports.py [hunk 7]
```diff
@@ -32,18 +33,48 @@ def load_items(self, filename):
...
+            # from Archive.org metadata, or provide json data book record
+            # which will be loaded directly into the OL catalog
+            record = lambda item: {'ia_id': item} if ia_items else {'data': json.dumps(item)}
+            values = [dict(batch_id=self.id, **record(item)) for item in items]
+            try:
+                # Upgrads psql and use `INSERT OR IGNORE`
+                # otherwise it will fail on UNIQUE `data`
+                # https://stackoverflow.com/questions/1009584
+                db.get_db().multiple_insert("import_item", values)
+            except:
```
Comment (mekarpeles) [inline openlibrary/core/imports.py:71]:
```suggestion
            except Exception:
```

File: openlibrary/core/imports.py [hunk 8]
```diff
@@ -32,18 +33,47 @@ def load_items(self, filename):
...
+            try:
+                # Upgrads psql and use `INSERT OR IGNORE`
+                # otherwise it will fail on UNIQUE `data`
+                # https://stackoverflow.com/questions/1009584
+                db.get_db().multiple_insert("import_item", values)
+            except Exception:
+                for value in values:
+                    try:
+                        db.get_db().insert("import_item", value)
+                    except Exception:
```
Comment (mekarpeles) [inline openlibrary/core/imports.py:76]:
```suggestion
                    except Exception:
                        logger.info("Record %s:%s failed to insert", self.name, value.id)
```

File: openlibrary/core/imports.py [hunk 9]
```diff
@@ -32,18 +33,48 @@ def load_items(self, filename):
...
+        ]
+        # ignore already present
+        logger.info(
+            "batch %s: %d items are already present, ignoring...",
+            self.name,
+            len(already_present)
+        )
+        return list(set(items) - set(already_present))
+
+    def add_items(self, items, ia_items=True):
```
Comment (cdrini) [inline openlibrary/core/imports.py:51]:
```suggestion
    def add_items(self, items, ia_items=True):
        """
        :param ia_items: Whether items is a list of IA identifiers. If False, expects a list of
        book data dicts.
        """
```

File: openlibrary/core/imports.py [hunk 10]
```diff
@@ -32,18 +33,48 @@ def load_items(self, filename):
...
+            values = [dict(batch_id=self.id, **record(item)) for item in items]
+            try:
+                # Upgrads psql and use `INSERT OR IGNORE`
+                # otherwise it will fail on UNIQUE `data`
+                # https://stackoverflow.com/questions/1009584
+                db.get_db().multiple_insert("import_item", values)
+            except Exception:
+                for value in values:
+                    try:
+                        db.get_db().insert("import_item", value)
```
Comment (mekarpeles) [inline openlibrary/core/imports.py:75]:
```suggestion
                        db.get_db().insert("import_item", **value)
```

File: openlibrary/core/imports.py [hunk 11]
```diff
@@ -32,18 +33,48 @@ def load_items(self, filename):
...
+            # which will be loaded directly into the OL catalog
+            def record(item):
+                return {'ia_id': item} if ia_items else {'data': json.dumps(item)}
+            values = [dict(batch_id=self.id, **record(item)) for item in items]
+            try:
+                # Upgrads psql and use `INSERT OR IGNORE`
+                # otherwise it will fail on UNIQUE `data`
+                # https://stackoverflow.com/questions/1009584
+                db.get_db().multiple_insert("import_item", values)
+            except Exception:
```
Comment (cdrini) [inline openlibrary/core/imports.py:72]: Pin down exception

File: openlibrary/core/imports.py [hunk 12]
```diff
@@ -5,6 +5,7 @@
 import datetime
 import time
 import web
+import json
 
 from psycopg2.errors import UndefinedTable
```
Comment (mekarpeles) [inline openlibrary/core/imports.py:10]:
```suggestion
from psycopg2.errors import UndefinedTable, UniqueViolation
```

File: openlibrary/core/imports.py (hunk 8 above)
Comment (mekarpeles) [reply openlibrary/core/imports.py:76]:
```suggestion
                    except UniqueViolation:
```

**Review by cdrini — APPROVED** [review]
Mek's en route with the changes. Code lgtm after these. We tested doing the import flow with BWB and looks like it's running smoothly!

Might want to do a test that IA is still running smoothly.

Merge at your discretion.

File: openlibrary/core/imports.py (hunk 11 above)
Comment (mekarpeles) [reply openlibrary/core/imports.py:80]:
```suggestion
            except UniqueViolation:
```

File: openlibrary/core/imports.py [hunk 13]
```diff
@@ -32,18 +33,56 @@ def load_items(self, filename):
...
+            # which will be loaded directly into the OL catalog
+            values = [
+                {
+                    'batch_id': self.id,
+                    **({'ia_id': item} if ia_items else {'data': json.dumps(item)})
+                }
+                for item in items
+            ]
+            try:
+                # Upgrads psql and use `INSERT OR IGNORE`
```
Comment (hornc) [inline openlibrary/core/imports.py:76]: Is this comment still necessary?  Ther's the dedupe method above, and the exception handling also - which one solves the problem, or are both needed?

File: openlibrary/core/imports.py [hunk 14]
```diff
@@ -63,7 +102,8 @@ def find_by_identifier(identifier):
             return ImportItem(result[0])
 
     def set_status(self, status, error=None, ol_key=None):
-        logger.info("set-status %s - %s %s %s", self.ia_id, status, error, ol_key)
+        _id = self.ia_id or "%s:%s" % (self.batch_id, self.id)
```
Comment (hornc) [inline openlibrary/core/imports.py:105]: This should be a trailing underscore, leading means private var, trailing is for avoiding name conflict by convention. https://hackernoon.com/understanding-the-underscore-of-python-309d1a029edc

File: openlibrary/core/imports.py (hunk 13 above)
Comment (mekarpeles) [reply openlibrary/core/imports.py:76]: 1. The dedupe method is only implemented for the `ia_id` field (not `data`)
2. The exception handling works, but ideally we'd upgrade psql and use the `insert or ignore` syntax

File: openlibrary/core/imports.py [hunk 15]
```diff
@@ -32,18 +33,56 @@ def load_items(self, filename):
...
+        logger.info(
+            "batch %s: %d items are already present, ignoring...",
+            self.name,
+            len(already_present)
+        )
+        return list(set(items) - set(already_present))
+
+    def add_items(self, items, ia_items=True):
+        """
+        :param ia_items: Whether items is a list of IA identifiers. If False, expects a list of
```
Comment (mekarpeles) [inline openlibrary/core/imports.py:53]:
```suggestion
        :param ia_items: True if `items` is a list of IA identifiers, False if
```

File: openlibrary/core/imports.py (hunk 13 above)
Comment (mekarpeles) [reply openlibrary/core/imports.py:76]:
```suggestion
                # TODO: Upgrade psql and use `INSERT OR IGNORE`
```

File: openlibrary/core/imports.py (hunk 14 above)
Comment (mekarpeles) [reply openlibrary/core/imports.py:105]:
```suggestion
        id_ = self.ia_id or "%s:%s" % (self.batch_id, self.id)
```

File: openlibrary/core/imports.py [hunk 16]
```diff
@@ -63,7 +102,8 @@ def find_by_identifier(identifier):
             return ImportItem(result[0])
 
     def set_status(self, status, error=None, ol_key=None):
-        logger.info("set-status %s - %s %s %s", self.ia_id, status, error, ol_key)
+        _id = self.ia_id or "%s:%s" % (self.batch_id, self.id)
+        logger.info("set-status %s - %s %s %s", _id, status, error, ol_key)
```
Comment (mekarpeles) [inline openlibrary/core/imports.py:106]:
```suggestion
        logger.info("set-status %s - %s %s %s", id_, status, error, ol_key)
```

**Comment by mekarpeles** [comment]
Tested w/ https://openlibrary.org/admin/imports/add

**Pull request merged**